### PR TITLE
Add helpers for auto-relinquishing shared resources used by a context

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**
Straight outta swarm.

### CODE REVIEW

File: src/ocean/util/container/VoidBufferAsArrayOf.d [hunk 1]
```diff
@@ -0,0 +1,220 @@
+/*******************************************************************************
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/VoidBufferAsArrayOf.d:1]: Not sure if `ocean.util.container` is a good location for this. Other buffery things live there, so put there by default.

File: src/ocean/util/container/VoidBufferAsArrayOf.d [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+*******************************************************************************/
+
+public struct VoidBufferAsArrayOf ( T )
+{
+    // T == void is not only pointless but is also invalid: it's illegal to pass
+    // a void argument to a function (e.g. opCatAssign).
+    static assert(!is(T == void));
+
+    /// Pointer to the unerlying void buffer. Must be set before use by struct
```
Comment (joseph-wakeling-sociomantic) [inline src/ocean/util/container/VoidBufferAsArrayOf.d]: `underlying` ;-)

File: src/ocean/util/container/pool/AcquiredResources.d [hunk 3]
```diff
@@ -0,0 +1,813 @@
...
+        T = type of resource
+
+*******************************************************************************/
+
+public struct Acquired ( T )
+{
+    import ocean.util.container.VoidBufferAsArrayOf;
+
+    /// Type of a new resource.
+    static if ( is(typeof({T* t = new T;})) )
```
Comment (joseph-wakeling-sociomantic) [inline src/ocean/util/container/pool/AcquiredResources.d]: Might be helpful to have a small comment to explain this `static if/else` choice ... ?  I presume it's meant to distinguish between cases where `T` is a value type and where `T` is a reference type?

**Comment by gavin-norman-sociomantic** [comment]
Updated with Joe's suggestions.

**Comment by gavin-norman-sociomantic** [comment]
I investigated adding a template to wrap a bunch of freelists and acquirers, based on a type tuple. However, it's complicated, as the `Acquired.acquire` method requires a `lazy Elem`, which would be awkward to specify. Maybe something for the future.

File: src/ocean/util/container/VoidBufferAsArrayOf.d (hunk 1 above)
Comment (mihails-strasuns) [reply src/ocean/util/container/VoidBufferAsArrayOf.d:1]: Makes sense to me.

**Review by mihails-strasuns — COMMENTED** [review]
This may need more simple examples in release notes. I can imagine that primary use case is to simply have bunch of different type buffers backed up by single `void[]` freelist. How does one achieve it?

**Comment by gavin-norman-sociomantic** [comment]
:+1: Makes sense. I'll add one.

**Comment by gavin-norman-sociomantic** [comment]
>I can imagine that primary use case is to simply have bunch of different type buffers backed up by single void[] freelist. How does one achieve it?

https://github.com/sociomantic-tsunami/ocean/pull/608/files#diff-60bf769b1fa6c58721c779b667ed7aa1R369 gives the basic idea of this. It's ~70 lines, though, so I don't think it makes sense to copy into the release notes. I've added a note about checking the usage examples in the module.

**Comment by gavin-norman-sociomantic** [comment]
Also added more usage info to the module header.

**Comment by mihails-strasuns** [comment]
Hm, it makes me wonder if this is of much use to apps in its current form. In 70 lines of code one can implement own wrapper for `FreeList` from scratch and I am not sure how many apps need resource management as convoluted as in swarm.

Don't mind merging this as-is but curious about your opinion - would it make sense to enhance this with something more simple or that would be better suited as a separate utility?

**Comment by gavin-norman-sociomantic** [comment]
Yeah, I wonder as well. We should get some of the people who were keen on having this in ocean to look at the PR and see if it would help them.

**Comment by gavin-norman-sociomantic** [comment]
(I suppose the actual definition of the shared resources class is only about 40 lines, in that example. The other lines are an example of using it in an execution context.)

>We should get some of the people who were keen on having this in ocean to look at the PR and see if it would help them.

@joseph-wakeling-sociomantic was one such person. Joe, the question is whether code like [this example](https://github.com/sociomantic-tsunami/ocean/blob/d0d49621b752c6198e421d62e5c7bf90eff5b0b9/src/ocean/util/container/pool/AcquiredResources.d#L392-L455) would be helpful to you, or whether it'd be easier to implement your own wrapper around `FreeList` in your app?

**Comment by joseph-wakeling-sociomantic** [comment]
I've got to be honest and say this all looks overcomplicated and intrusive to me.  It looks like it works well for a case where (say) one is grabbing a bunch of temporary resources at the start of a `Task` context, or maybe where one has a resource supplier as part of a context object.  It's possibly not so useful for the use-cases I have in mind where the code is generally a class that encapsulates some calculations requiring temporary buffers under the hood.

What would probably be more useful for my case is a _thin_ wrapper around `FreeList` that allows for the option to get and recycle buffers with different element types.  Typical use case would be something like:

```D
class MyMagicClass
{
    // privately encapsulated supplier of buffers
    private MagicBufferSupplier buffer_pool;

    this () { this.buffer_pool = new MagicBufferSupplier; }

    // templated method, so this can be invoked for different data types
    // and Just Work with one underlying buffer pool
    void foo (T) (...)
    {
        auto buf = this.buffer_pool.get!(T)(...);
        scope (exit) this.buffer_pool.recycle(buf);

        // ... do stuff with the buffer
    }
}
```

... makes sense?

**Comment by joseph-wakeling-sociomantic** [comment]
Of course, note that part of the reason behind the above is a wish to encapsulate the buffer supplier.  This makes sense where e.g. the code written is not as part of an app but as a standalone library component (which is a long-term design goal for the particular code I have in mind).  I imagine that for app components it would be more reasonable to have the resource supplier as a global app resource.

**Comment by gavin-norman-sociomantic** [comment]
Yeah, what you have in mind is a bit different. Could probably be implemented on top of what's already in this PR, though.

**Comment by joseph-wakeling-sociomantic** [comment]
Can you talk through a bit how you see that working?  It's not obvious to me how it could operate without exposing app or library functionality to a large chunk of global mutable state.  (But maybe you're not seeing that as a problem?)

**Comment by gavin-norman-sociomantic** [comment]
Perhaps I'm missing something. There's nothing inherent in the code added by this PR that means that the shared resources must be global. You could, for example, instantiate a shared resources pool inside `MyMagicClass`, to use your example.

**Comment by joseph-wakeling-sociomantic** [comment]
I guess so.  It just feels like the amount of boilerplate to do so is quite large (but maybe I'm missing something too).  I'd assumed that the typical use-case for it would be as one resource for an entire app.

**Comment by gavin-norman-sociomantic** [comment]
Agreed. I just wanted to clarify that we're on the same page re: it not being intrinsically global.

So the main point of this PR is to provide building blocks for:
* A shared resources pool. (Intended to be owned at the app level, but that's not mandatory.)
* A tracker for resources acquired by a task (or other context), with the ability to automatically relinquish them back to the shared pool when the task ends.

Are either of those points applicable to your use case? (I'm getting the feeling not.)

**Comment by joseph-wakeling-sociomantic** [comment]
> Are either of those points applicable to your use case?

Not really, no.  I think the use-case is much more of small encapsulated buffer pools for specific pieces of functionality, where relinquishing the resources is not tied to `Task` completion but (typically) to scope exit.

The one thing that _is_ needed for my use-case is the ability to have one pool which can serve multiple different types of buffer from the caller's point of view.

**Comment by gavin-norman-sociomantic** [comment]
>where relinquishing the resources is not tied to Task completion but (typically) to scope exit.

The exact scope at which the resources are relinquished isn't important. (My mention of `Task`s was just an example. In fact, in *proto, we relinquish on `scope(exit)`, as you say.)

**Comment by joseph-wakeling-sociomantic** [comment]
OK.  So sounds like the second point is a common area of interest.  From a usage point of view it's convenient if the code is not boilerplate-y, i.e. it doesn't require up-front declaration of the different supported types (cf. my pseudo-code example above).

**Comment by gavin-norman-sociomantic** [comment]
Yep. I could add a simplistic buffer getter built on top of the existing code in this PR, say.

**Comment by gavin-norman-sociomantic** [comment]
Usage would look like this:
```D
class MyMagicClass
{
    private MagicBufferSupplier buffer_pool;

    this () { this.buffer_pool = new MagicBufferSupplier; }

    // templated method, so this can be invoked for different data types
    // and Just Work with one underlying buffer pool
    void foo (T) (...)
    {
        scope buffers = this.buffer_pool.new AcquiredResources;

        // get one or more buffers in this context.
        // they will be automatically relinquished on scope(exit)
        buffers.get!(T)(...);

        // ... do stuff with the acquired buffers
    }
}
```

**Comment by joseph-wakeling-sociomantic** [comment]
Seems reasonable.  A bit of a shame the boilerplate-y `scope buffers = ...` line is needed, but that's no big deal.

Will the individual buffers be recycled when the scope of the `AcquiredResources` instance ends, or when their own scope ends?  For the use-case I have there is some convenience in being able to recycle individual buffers at the end of their scope.

**Comment by gavin-norman-sociomantic** [comment]
>Seems reasonable. A bit of a shame the boilerplate-y scope buffers = ... line is needed, but that's no big deal.

But you don't have to manually relinquish the buffers. So that line from your version is removed ;)

>Will the individual buffers be recycled when the scope of the AcquiredResources instance ends, or when their own scope ends?

When `AcquiredResources`' scope exits. If you want sub-scopes, you can add them manually with `{}`, as usual.

**Comment by joseph-wakeling-sociomantic** [comment]
OK, all good then.  There might be some value in allowing some generic "get and recycle a buffer with element type `T`" functionality, versus a helper class like `AcquiredResources` that manages the scoping on your behalf -- but that's not a really big deal.  This seems like a nice addition to ocean.

**Comment by gavin-norman-sociomantic** [comment]
Yeah, it sounds to me like what you really want, Joe, is a simple wrapper around a free-list that allows one to acquire and recycle arbitrarily typed buffers (without any scoped tracking or auto-relinquishing mechanism). This is a different thing, really.

**Comment by joseph-wakeling-sociomantic** [comment]
Agreed.  And that is what I noted right at the beginning of this discussion ;-)
https://github.com/sociomantic-tsunami/ocean/pull/608#issuecomment-415359907

**Comment by gavin-norman-sociomantic** [comment]
>We should get some of the people who were keen on having this in ocean to look at the PR and see if it would help them.

So it seems Joe isn't really the intended audience ;) We should find someone who is. (Or perhaps the whole idea of this PR is specific to swarm and rumours of people needing such support in ocean were ill-founded!)

**Comment by joseph-wakeling-sociomantic** [comment]
Personally I think it's a useful feature for ocean, it's just not appropriate to the one particular use-case that sparked the discussion.

I can almost certainly find other contexts in which the use-case is valid.  And it's not like it's _invalid_ for my actual use-case; it's just that other tradeoffs mean that it may be preferable to approach that from a slightly different point of view.

**Comment by gavin-norman-sociomantic** [comment]
I can confirm that two applications use exactly the pattern that this PR implements. (Albeit with very simplistic use cases, where only buffers of a single type are acquired.)

**Comment by nemanja-boric-sociomantic** [comment]
There are also more advanced requirements - such as acquiring the buffers based on the anticipated grow size, and having a rudimentary support in the ocean is a good step forward towards implementing more advanced memory management patterns.

**Comment by joseph-wakeling-sociomantic** [comment]
Yup. If you look at my use-case (which involves creating buffers to interact with C number-crunching libraries), a need is to be able to specify a _minimum_ length for the buffer.  In this way we avoid any need for costly `enableStomping`/`assumeSafeAppend` calls.

**Comment by joseph-wakeling-sociomantic** [comment]
If we had those sort of extra helper features, it would make a lot more sense to use `AcquiredResources` etc.

**Comment by mihails-strasuns** [comment]
@gavin-norman-sociomantic can you sum things up regarding what to do here? :)

**Comment by gavin-norman-sociomantic** [comment]
Summary:
* The PR as presented is heavily used in swarm / *proto.
* Other real-time server apps in our system could use the PR as presented, but only its basic features (acquiring and relinquishing buffers from/to a pool).
* What Joe wants is something a bit different. (But could be implemented on top of what's in this PR.)

**Comment by mihails-strasuns** [comment]
== let's merge and make another PR for Joe utils later?

**Comment by gavin-norman-sociomantic** [comment]
I feel that would be reasonable.

**Comment by joseph-wakeling-sociomantic** [comment]
I'm fine with that.

**Review by mihails-strasuns — APPROVED** [review]

**Comment by gavin-norman-sociomantic** [comment]
@joseph-wakeling-sociomantic care to create an issue explaining what features you'd like?

**Comment by joseph-wakeling-sociomantic** [comment]
Sure thing.  I'll follow up later today.

**Pull request merged**